File: sdk/python/scripts.py
"""
脚本管理和版本下载模块
支持脚本版本检查、下载和版本更新

功能特性：
- 获取脚本版本信息
- 下载脚本文件
- 检查脚本更新
- 下载版本发布文件

使用示例：
    from license_client import LicenseClient
    from scripts import ScriptManager, ReleaseManager

    # 初始化
    client = LicenseClient(server_url, app_key, skip_verify=True)

    # 脚本管理
    script_manager = ScriptManager(client)
    versions = script_manager.get_script_versions()
    content = script_manager.download_script("script.py")

    # 版本下载
    release_manager = ReleaseManager(client)
    release_manager.download_release("app_v1.0.0.zip", "./downloads/app.zip")
"""

import os
from typing import Optional, Dict, List, Callable, Tuple
from dataclasses import dataclass
from urllib.parse import quote


@dataclass
class ScriptInfo:
    """脚本信息"""
    filename: str
    version: str
    version_code: int
    file_size: int
    file_hash: str
    updated_at: str


@dataclass
class ScriptVersionResponse:
    """脚本版本响应"""
    scripts: List[ScriptInfo]
    total_count: int
    last_updated: str


@dataclass
class UpdateInfo:
    """更新信息"""
    version: str
    version_code: int
    download_url: str
    changelog: str
    file_size: int
    file_hash: str
    force_update: bool


class ScriptManager:
    """脚本管理器"""

    def __init__(self, license_client):
        """
        初始化脚本管理器

        Args:
            license_client: LicenseClient 实例
        """
        self.client = license_client

    def get_script_versions(self) -> ScriptVersionResponse:
        """
        获取脚本版本信息

        Returns:
            脚本版本响应，包含所有可用脚本的版本信息
        """
        url = f"{self.client.server_url}/api/client/scripts/version"
        params = {"app_key": self.client.app_key}

        resp = self.client._session.get(url, params=params, timeout=self.client.timeout)
        result = resp.json()

        if result.get('code') != 0:
            raise Exception(result.get('message', '获取脚本版本失败'))

        data = result.get('data', {})
        scripts = [ScriptInfo(
            filename=s.get('filename', ''),
            version=s.get('version', ''),
            version_code=s.get('version_code', 0),
            file_size=s.get('file_size', 0),
            file_hash=s.get('file_hash', ''),
            updated_at=s.get('updated_at', '')
        ) for s in data.get('scripts', [])]

        return ScriptVersionResponse(
            scripts=scripts,
            total_count=data.get('total_count', len(scripts)),
            last_updated=data.get('last_updated', '')
        )

    def download_script(self, filename: str, save_path: Optional[str] = None) -> bytes:
        """
        下载指定脚本文件

        Args:
            filename: 脚本文件名
            save_path: 保存路径（如果为空，返回内容而不保存）

        Returns:
            脚本内容
        """
        url = f"{self.client.server_url}/api/client/scripts/{quote(filename)}"
        params = {
            "app_key": self.client.app_key,
            "machine_id": self.client.machine_id
        }

        resp = self.client._session.get(url, params=params, timeout=self.client.timeout)

        if resp.status_code != 200:
            try:
                result = resp.json()
                raise Exception(result.get('message', f'下载失败: HTTP {resp.status_code}'))
            except:
                raise Exception(f'下载失败: HTTP {resp.status_code}')

        content = resp.content

        # 如果指定了保存路径，保存到文件
        if save_path:
            # 确保目录存在
            dir_path = os.path.dirname(save_path)
            if dir_path:
                os.makedirs(dir_path, exist_ok=True)

            with open(save_path, 'wb') as f:
                f.write(content)

        return content

    def check_script_update(self, filename: str, current_version_code: int) -> Tuple[bool, Optional[ScriptInfo]]:
        """
        检查脚本是否有更新

        Args:
            filename: 脚本文件名
            current_version_code: 当前版本号

        Returns:
            (是否有更新, 最新版本信息)
        """
        versions = self.get_script_versions()

        for script in versions.scripts:
            if script.filename == filename:
                if script.version_code > current_version_code:
                    return True, script
                return False, script

        raise Exception(f"脚本 {filename} 不存在")


class ReleaseManager:
    """版本发布管理器"""

    def __init__(self, license_client):
        """
        初始化版本发布管理器

        Args:
            license_client: LicenseClient 实例
        """
        self.client = license_client

    def download_release(
        self,
        filename: str,
        save_path: str,
        progress_callback: Optional[Callable[[int, int], None]] = None
    ) -> None:
        """
        下载版本文件

        Args:
            filename: 文件名
            save_path: 保存路径
            progress_callback: 下载进度回调 (已下载字节数, 总字节数)
        """
        url = f"{self.client.server_url}/api/client/releases/download/{quote(filename)}"
        params = {
            "app_key": self.client.app_key,
            "machine_id": self.client.machine_id
        }

        resp = self.client._session.get(url, params=params, stream=True, timeout=self.client.timeout)

        if resp.status_code != 200:
            try:
                result = resp.json()
                raise Exception(result.get('message', f'下载失败: HTTP {resp.status_code}'))
            except:
                raise Exception(f'下载失败: HTTP {resp.status_code}')

        # 确保目录存在
        dir_path = os.path.dirname(save_path)
        if dir_path:
            os.makedirs(dir_path, exist_ok=True)

        # 获取文件大小
        total_size = int(resp.headers.get('content-length', 0))

        # 下载文件
        with open(save_path, 'wb') as f:
            downloaded = 0
            for chunk in resp.iter_content(chunk_size=32 * 1024):
                if chunk:
                    f.write(chunk)
                    downloaded += len(chunk)
                    if progress_callback and total_size > 0:
                        progress_callback(downloaded, total_size)

    def get_latest_release_and_download(
        self,
        save_path: str,
        progress_callback: Optional[Callable[[int, int], None]] = None
    ) -> UpdateInfo:
        """
        获取最新版本并下载

        Args:
            save_path: 保存路径
            progress_callback: 下载进度回调

        Returns:
            更新信息
        """
        # 获取最新版本信息
        update_info = self.client.check_update()
        if not update_info:
            raise Exception("没有可用的更新")

        # 从 download_url 提取文件名
        download_url = update_info.get('download_url', '')
        if not download_url:
            raise Exception("无效的下载URL")

        filename = os.path.basename(download_url)
        if not filename or filename == '.':
            raise Exception("无效的下载URL")

        # 下载文件
        self.download_release(filename, save_path, progress_callback)

        return UpdateInfo(
            version=update_info.get('version', ''),
            version_code=update_info.get('version_code', 0),
            download_url=download_url,
            changelog=update_info.get('changelog', ''),
            file_size=update_info.get('file_size', 0),
            file_hash=update_info.get('file_hash', ''),
            force_update=update_info.get('force_update', False)
        )

File: sdk/go/scripts.go
// Package license 脚本管理和版本下载功能
package license

import (
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"os"
	"path/filepath"
)

// ScriptInfo 脚本信息
type ScriptInfo struct {
	Filename    string `json:"filename"`
	Version     string `json:"version"`
	VersionCode int    `json:"version_code"`
	FileSize    int64  `json:"file_size"`
	FileHash    string `json:"file_hash"`
	UpdatedAt   string `json:"updated_at"`
}

// ScriptVersionResponse 脚本版本响应
type ScriptVersionResponse struct {
	Scripts     []ScriptInfo `json:"scripts"`
	TotalCount  int          `json:"total_count"`
	LastUpdated string       `json:"last_updated"`
}

// ReleaseDownloadInfo 版本下载信息
type ReleaseDownloadInfo struct {
	Filename    string `json:"filename"`
	DownloadURL string `json:"download_url"`
	FileSize    int64  `json:"file_size"`
	FileHash    string `json:"file_hash"`
}

// ScriptManager 脚本管理器
type ScriptManager struct {
	client *Client
}

// NewScriptManager 创建脚本管理器
func (c *Client) NewScriptManager() *ScriptManager {
	return &ScriptManager{
		client: c,
	}
}

// GetScriptVersions 获取脚本版本信息
// 返回所有可用脚本的版本信息列表
func (m *ScriptManager) GetScriptVersions() (*ScriptVersionResponse, error) {
	params := url.Values{}
	params.Set("app_key", m.client.appKey)

	resp, err := m.client.httpClient.Get(m.client.serverURL + "/api/client/scripts/version?" + params.Encode())
	if err != nil {
		return nil, fmt.Errorf("请求失败: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("读取响应失败: %w", err)
	}

	var result struct {
		Code    int                   `json:"code"`
		Message string                `json:"message"`
		Data    ScriptVersionResponse `json:"data"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("解析响应失败: %w", err)
	}

	if result.Code != 0 {
		return nil, fmt.Errorf("API错误: %s", result.Message)
	}

	return &result.Data, nil
}

// DownloadScript 下载指定脚本文件
// filename: 脚本文件名
// savePath: 保存路径（如果为空，返回内容而不保存）
func (m *ScriptManager) DownloadScript(filename string, savePath string) ([]byte, error) {
	params := url.Values{}
	params.Set("app_key", m.client.appKey)
	params.Set("machine_id", m.client.machineID)

	downloadURL := fmt.Sprintf("%s/api/client/scripts/%s?%s",
		m.client.serverURL, url.PathEscape(filename), params.Encode())

	resp, err := m.client.httpClient.Get(downloadURL)
	if err != nil {
		return nil, fmt.Errorf("下载失败: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		var errResult struct {
			Code    int    `json:"code"`
			Message string `json:"message"`
		}
		if json.Unmarshal(body, &errResult) == nil && errResult.Message != "" {
			return nil, fmt.Errorf("下载失败: %s", errResult.Message)
		}
		return nil, fmt.Errorf("下载失败: HTTP %d", resp.StatusCode)
	}

	content, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("读取内容失败: %w", err)
	}

	// 如果指定了保存路径，保存到文件
	if savePath != "" {
		// 确保目录存在
		dir := filepath.Dir(savePath)
		if err := os.MkdirAll(dir, 0755); err != nil {
			return nil, fmt.Errorf("创建目录失败: %w", err)
		}

		if err := os.WriteFile(savePath, content, 0644); err != nil {
			return nil, fmt.Errorf("保存文件失败: %w", err)
		}
	}

	return content, nil
}

// CheckScriptUpdate 检查脚本是否有更新
// filename: 脚本文件名
// currentVersion: 当前版本号
// 返回: 是否有更新, 最新版本信息, 错误
func (m *ScriptManager) CheckScriptUpdate(filename string, currentVersionCode int) (bool, *ScriptInfo, error) {
	versions, err := m.GetScriptVersions()
	if err != nil {
		return false, nil, err
	}

	for _, script := range versions.Scripts {
		if script.Filename == filename {
			if script.VersionCode > currentVersionCode {
				return true, &script, nil
			}
			return false, &script, nil
		}
	}

	return false, nil, fmt.Errorf("脚本 %s 不存在", filename)
}

// ReleaseManager 版本发布管理器
type ReleaseManager struct {
	client *Client
}

// NewReleaseManager 创建版本发布管理器
func (c *Client) NewReleaseManager() *ReleaseManager {
	return &ReleaseManager{
		client: c,
	}
}

// DownloadRelease 下载版本文件
// filename: 文件名
// savePath: 保存路径
// progressCallback: 下载进度回调 (已下载字节数, 总字节数)
func (m *ReleaseManager) DownloadRelease(filename string, savePath string, progressCallback func(downloaded, total int64)) error {
	params := url.Values{}
	params.Set("app_key", m.client.appKey)
	params.Set("machine_id", m.client.machineID)

	downloadURL := fmt.Sprintf("%s/api/client/releases/download/%s?%s",
		m.client.serverURL, url.PathEscape(filename), params.Encode())

	resp, err := m.client.httpClient.Get(downloadURL)
	if err != nil {
		return fmt.Errorf("下载失败: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		var errResult struct {
			Code    int    `json:"code"`
			Message string `json:"message"`
		}
		if json.Unmarshal(body, &errResult) == nil && errResult.Message != "" {
			return fmt.Errorf("下载失败: %s", errResult.Message)
		}
		return fmt.Errorf("下载失败: HTTP %d", resp.StatusCode)
	}

	// 确保目录存在
	dir := filepath.Dir(savePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("创建目录失败: %w", err)
	}

	// 创建文件
	file, err := os.Create(savePath)
	if err != nil {
		return fmt.Errorf("创建文件失败: %w", err)
	}
	defer file.Close()

	// 获取文件大小
	totalSize := resp.ContentLength

	// 如果有进度回调，使用带进度的复制
	if progressCallback != nil && totalSize > 0 {
		var downloaded int64
		buf := make([]byte, 32*1024) // 32KB buffer
		for {
			n, err := resp.Body.Read(buf)
			if n > 0 {
				_, writeErr := file.Write(buf[:n])
				if writeErr != nil {
					return fmt.Errorf("写入文件失败: %w", writeErr)
				}
				downloaded += int64(n)
				progressCallback(downloaded, totalSize)
			}
			if err == io.EOF {
				break
			}
			if err != nil {
				return fmt.Errorf("读取数据失败: %w", err)
			}
		}
	} else {
		// 直接复制
		if _, err := io.Copy(file, resp.Body); err != nil {
			return fmt.Errorf("保存文件失败: %w", err)
		}
	}

	return nil
}

// GetLatestReleaseAndDownload 获取最新版本并下载
// savePath: 保存路径
// progressCallback: 下载进度回调
func (m *ReleaseManager) GetLatestReleaseAndDownload(savePath string, progressCallback func(downloaded, total int64)) (*UpdateInfo, error) {
	// 获取最新版本信息
	updateInfo, err := m.client.CheckUpdate()
	if err != nil {
		return nil, fmt.Errorf("获取版本信息失败: %w", err)
	}

	if updateInfo == nil || updateInfo.DownloadURL == "" {
		return nil, fmt.Errorf("没有可用的更新")
	}

	// 从 DownloadURL 提取文件名
	filename := filepath.Base(updateInfo.DownloadURL)
	if filename == "" || filename == "." {
		return nil, fmt.Errorf("无效的下载URL")
	}

	// 下载文件
	if err := m.DownloadRelease(filename, savePath, progressCallback); err != nil {
		return nil, err
	}

	return updateInfo, nil
}
